### PR TITLE
Reinstate the KDF error macros

### LINKED ISSUE #10340: kdferr.h got removed from installed headers
Comparing the list of installed header files from 1.1.1 to current master, I notice at least kdferr.h got removed.

We also have the following new public files, not sure if they're all supposed to be public:
cmperr.h
cmp.h
cmp_util.h
core.h
core_names.h
core_numbers.h
crmferr.h
crmf.h
esserr.h
ess.h
fips_names.h
macros.h
params.h
provider.h
trace.h
types.h

### PR DESCRIPTION
**Pull request by levitte**
include/openssl/kdferr.h was removed because it got empty.  However,
that will break the build of any application that includes it
directly, so we add it back as a placeholder.

Fixes #10340

### CODE REVIEW

**Review by t-j-h — DISMISSED** [review]

**Comment by kroeckx** [comment]
I expect all the old defines to be in there. That is, the KDF_F_* defined to 0, the KDF_R_* to some distinct value >= 100. There also was an ERR_load_KDF_strings().

**Review by kroeckx — CHANGES_REQUESTED** [review]
All old defines/functions should be in it.

**Comment by levitte** [comment]
Ah, so what you're asking for is a revert of fe6ec26b204a056aee2a24b79df09a45b2308603, basically

**Review by kroeckx — APPROVED** [review]

**Comment by kroeckx** [comment]
# Number of lines in @nm_lines before massaging: 4935
# Number of lines in @def_lines before massaging: 4795
# Number of lines in @nm_lines after massaging: 4790
# Number of lines in @def_lines after massaging: 4791
# The following symbols are missing in libcrypto.so:
#   ERR_load_KDF_strings

File: include/openssl/kdferr.h [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+# define HEADER_KDFERR_H
+
+# include <openssl/opensslconf.h>
+# include <openssl/symhacks.h>
+
+
+# ifdef  __cplusplus
+extern "C"
+# endif
+int ERR_load_KDF_strings(void);
```
Comment (kroeckx) [inline include/openssl/kdferr.h]: I guess you want to mark this as deprecated

**Review by kroeckx — CHANGES_REQUESTED** [review]

**Comment by kroeckx** [comment]
and run make update

**Comment by levitte** [comment]
Ah, yeah, forgot that.

File: include/openssl/kdferr.h (hunk 1 above)
Comment (levitte) [reply include/openssl/kdferr.h]: Good idea

**Comment by levitte** [comment]
I reverted a lot more than might be practically needed...  I could see reason to empty `crypto/kdf/kdf_err.c` of almost all content, just leaving `ERR_load_KDF_strings()` as a function that does nothing.

Still, it was an interesting exercise, to invent some kind of "conservation state" for this...  we may need to do the same with others, eventually.

**Comment by kroeckx** [comment]
I don't see a reason to keep the strings in kdf_err.c

**Comment by kroeckx** [comment]
```
crypto: function ERR_load_KDF_strings undocumented
# 1 in util/libcrypto.num are not documented
```

**Comment by levitte** [comment]
Eh gods. Could we perhaps make a documentation exception on deprecated functions?
Meanwhile, I'll add it to the appropriate missing file.

**Comment by levitte** [comment]
All done

File: crypto/err/err_all.c [hunk 2]
```diff
@@ -38,6 +38,7 @@
 #include <openssl/cmperr.h>
 #include <openssl/cterr.h>
 #include <openssl/asyncerr.h>
+#include <openssl/kdferr.h>
```
Comment (kroeckx) [inline crypto/err/err_all.c]: Why this include?

File: include/openssl/kdferr.h [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+#  define KDF_F_X942KDF_HASH_KDM                           0
+#  define KDF_F_X942KDF_NEW                                0
+#  define KDF_F_X942KDF_SIZE                               0
+#  define KDF_F_X963KDF_DERIVE                             0
+# endif
+
+/*
+ * KDF reason codes.
+ */
+# define KDF_R_BAD_ENCODING                               122
```
Comment (kroeckx) [inline include/openssl/kdferr.h]: Mark as deprecated too?

File: crypto/err/err_all.c (hunk 2 above)
Comment (levitte) [reply crypto/err/err_all.c]: Because I actually did a revert and didn't notice this :wink:

File: include/openssl/kdferr.h (hunk 3 above)
Comment (levitte) [reply include/openssl/kdferr.h]: Good idea

File: crypto/err/err_all.c (hunk 2 above)
Comment (levitte) [reply crypto/err/err_all.c]: Fixed

File: include/openssl/kdferr.h (hunk 3 above)
Comment (levitte) [reply include/openssl/kdferr.h]: Done

**Comment by kroeckx** [comment]
I think it looks good, but I'm not going to review (the perl part).

**Review by t8m — APPROVED** [review]
Looks good, including the perl.

**Comment by levitte** [comment]
Merged.

7c6a0d909abda7c854b44831ef43ebdda49ed42e Reinstate the KDF error macros
dc5ce51983713dfbb8f3d0c8b198dee32b563ac6 util/perl/OpenSSL/ParseC.pm: Fix 'extern "C" DEPRECATEDIN_x_y(...)'
905b097fd586ed0279aee05b5961224212014ec9 Deprecate ERR_load_KDF_strings()
4d301427a96010468da2bb67bf1025fa8d886ab9 Make sure KDF reason codes are conserved in their current state

**Pull request closed without merging**